juju_verify/verifiers/neutron_gateway.py

```diff
@@ -0,0 +1,180 @@
+
+    NAME = 'neutron-gateway'
+    action_name_result_map = {"get-status-routers": "router-list",
+                              "get-status-dhcp": "dhcp-networks",
+                              "get-status-lb": "load-balancers"}
+    action_name_failure_string_map = {"get-status-routers": ("The following routers are "
+                                                             "non-redundant: {}"),
+                                      "get-status-dhcp": ("The following DHCP networks "
+                                                          "are non-redundant: {}"),
+                                      "get-status-lb": ("The following LBaasV2 LBs were "
+                                                        "found: {}. LBaasV2 does not "
+                                                        "offer HA.")}
+
+    def __init__(self, units: List[Unit]) -> None:
+        """Neutron Gateway verifier constructor."""
+        super().__init__(units)
+        self.cache_action_resource_list_map: Dict[str, List] = {}
+
+    def get_all_ngw_units(self) -> List[Unit]:
+        """Get all neutron-gateway units, including those not being shutdown."""
+        all_ngw_units = []
+        for unit in self.model.units.values():
+            charm = parse_charm_name(unit.data.get('charm-url', ''))
+            if charm == self.NAME:
+                all_ngw_units.append(unit)
+        return all_ngw_units
+
+    def get_resource_list(self, get_resource_action_name: str) -> List[dict]:
+        """Given a get resource action, return matching resources from all units."""
+        try:
+            return self.cache_action_resource_list_map[get_resource_action_name]
+        except KeyError:
+            pass
+
+        resource_list = []
+        shutdown_hostname_list = [get_unit_hostname(unit) for unit in self.units]
+
+        for unit in self.get_all_ngw_units():
+            hostname = get_unit_hostname(unit)
+            host_resource_list = get_unit_resource_list(unit,
+                                                        get_resource_action_name)
+
+            # add host metadata to resource
+            for resource in host_resource_list:
+                resource["host"] = hostname
+                resource["juju-entity-id"] = unit.entity_id
+                resource["shutdown"] = False
+
+                if hostname in shutdown_hostname_list:
+                    resource["shutdown"] = True
+
+                resource_list.append(resource)
+
+        self.cache_action_resource_list_map[get_resource_action_name] = resource_list
+        return resource_list
+
+    def get_shutdown_resource_list(self, get_resource_action_name: str) -> List[dict]:
+        """Return a list of resources matching action that are going to be shutdown."""
+        res_list = self.get_resource_list(get_resource_action_name)
+        return [r for r in res_list if r["shutdown"] and r["status"] == "ACTIVE"]
```

Comment:
I personally rather avoid using single-letter name for the variable. In this case, the function is really short and understandable, but even so, I would probably prefer not to use one-letter variables.

Reply:
I'm surprised that neither `pylint` nor `flake8` complained about the single letter variables. Good point.